### PR TITLE
Destructible Automatic Overkill

### PR DESCRIPTION
**Pull request by kontakt**
## About the PR

- Added an automatic overkill behavior threshold for entities with flat damage destruction thresholds.
  - Test for same.
- Added a method to add thresholds to an existing destructible component.
- Cleaned up destructible somewhat.
- Removed manual destruction from `LockerBase` as a demo.

## Why / Balance

Adding automatic overkill behavior reduces YAML boilerplate for destructible entities. 
Thresholds are set to `Max(100, 2 * maxTrigger)`, resulting in consistent scaling for entities and ensuring overkill thresholds should not be triggered by random non-explosive events. 

Adding thresholds and associated acts is desired by other systems in the codebase, e.g. staff of animation needs to add destruction behavior for items that already have destructible but lack a destruction node like fragile mail.

Cleanup of namespaces and comments is good practice to slowly improve the codebase.

## Technical details

New method: `public void AddThreshold(Entity<DestructibleComponent?> entity, DamageThreshold threshold, int? index = null)`
New method: `private void AddOverkillThreshold(Entity<DestructibleComponent> entity)`
New method: `private void OnMapInit(EntityUid uid, DestructibleComponent component, MapInitEvent args)`

On MapInit, `AddOverkillThreshold` is called and iterates over the thresholds and behaviors to identify if a destruction act is reachable by nonspecific damage. If one is found that is not _solely_ destruction (a signature of manual overkill behavior), an overkill threshold is added at the head of the list.

An `IgnoreExcess` member is added to DestructibleComponent to disable automatic overkill if necessary. 

Observe an example of the YAML savings in `LockerBase`, six lines removed. This pattern is visible in many locations in the codebase. A followup PR will remove these instances once this is merged.

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an in-game showcase.

## Breaking changes

Destructible now has automatic overkill behavior. Any prototypes with a destruction act that must occur even in cases of extreme damage (i.e. the nuke) need to add `ignoreExcess` to their destructible component. 

**Changelog**

No (major) player visible changes.

### CODE REVIEW

File: Content.Server/Destructible/DestructibleComponent.cs [hunk 1]
```diff
@@ -1,25 +1,30 @@
...
-        /// Specifies whether the entity has passed a damage threshold that causes it to break
-        /// </summary>
-        [DataField]
-        public bool IsBroken = false;
-    }
+    /// <summary>
+    /// Specifies whether the entity should ignore excess damage and handle destruction normally.
+    /// </summary>
+    [DataField(readOnly: true)]
+    public bool IgnoreExcess = false;
```
Comment (moonheart08) [inline Content.Server/Destructible/DestructibleComponent.cs]: Nit: I'd invert this, make it `GenerateOverkillThreshhold` or similar to better describe what it does.

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 2]
```diff
@@ -19,210 +18,293 @@
...
-        {
-            component.IsBroken = false;
+    /// <summary>
+    /// Map Initialization function for <see cref="DestructibleComponent"/>, adding automatic overkill threshold.
+    /// </summary>
+    /// <param name="entity">The uid, component tuple.</param>
+    /// <param name="args">The event arguments.</param>
+    private void OnMapInit(Entity<DestructibleComponent> entity, ref MapInitEvent args)
+    {
+        AddOverkillThreshold(entity);
```
Comment (moonheart08) [inline Content.Server/Destructible/DestructibleSystem.cs:69]: This is setting up component data in a way that only relies on the component's own fields, and should be on ComponentInit (not startup, which happens later). This ensures it runs early for any startup-time or later logic, extra important as the field is `readOnly` and will thus never be saved.

The logic appears to correctly handle having already run, so putting it on init should pose no issue.

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 3]
```diff
@@ -19,210 +18,293 @@
...
+    /// <param name="entity">The entity, component tuple to target.</param>
+    /// <param name="threshold">The threshold to add.</param>
+    /// <param name="index">The index at which to insert the threshold.</param>
+    public void AddThreshold(Entity<DestructibleComponent?> entity, DamageThreshold threshold, int? index = null)
+    {
+        if (!Resolve(entity.Owner, ref entity.Comp, false))
+            entity.Comp = AddComp<DestructibleComponent>(entity.Owner);
+
+        if(index is not null)
+            entity.Comp.Thresholds.Insert(Math.Clamp(index.Value, 0, entity.Comp.Thresholds.Count), threshold);
```
Comment (moonheart08) [inline Content.Server/Destructible/DestructibleSystem.cs]: Bad index should throw, additionally this function should take an https://learn.microsoft.com/en-us/dotnet/api/system.index?view=net-10.0 instead. Index has an implicit cast from integer, so the API is preserved.

This will allow indexing the threshhold list from the end, though you'll need to support that manually in here by computing the integer index from the Index. (Maybe make a helper for this in Content.Shared/Utility/, `IndexExtensions`, that does this for you for a given index and collection length?)

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 4]
```diff
@@ -19,210 +18,293 @@
...
+                    if (logImpact <= behavior.Impact)
+                        logImpact = behavior.Impact;
+                    if (behavior is DoActsBehavior doActsBehavior)
                     {
-                        AdminLogger.Add(LogType.Damaged,
-                            logImpact,
-                            $"Unknown damage source caused {ToPrettyString(uid):subject} to trigger [{triggeredBehaviors}]");
+                        return $"{behavior.GetType().Name}:{doActsBehavior.Acts.ToString()}";
                     }
+                    return behavior.GetType().Name;
```
Comment (moonheart08) [inline Content.Server/Destructible/DestructibleSystem.cs:97]: Refactoring remark: This should probably be done with an interface down the line so there isn't hardcoded logic in here. Not required for this PR.

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+            var threshold = sTestThresholdListenerSystem.ThresholdsReached[0].Threshold;
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(threshold.Triggered, Is.True);
+                Assert.That(threshold.Behaviors, Has.Count.EqualTo(1));
+            });
+
+            var entitiesInRange = sEntityManager.System<EntityLookupSystem>().GetEntitiesInRange(coordinates, 3, LookupFlags.All | LookupFlags.Approximate);
```
Comment (moonheart08) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Your destructible entity should be the only one in existence during the test. You don't need a range lookup in an empty map that only has a grid and whatever you spawned.

**Review by moonheart08 — CHANGES_REQUESTED** [review]

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 5 above)
Comment (Tayrtahn) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: I suspect the intention is to make sure that the overkill destruction doesn't spawn any new entities. However, this check does seem needlessly complicated - I would just check the entity count from `EntityManager`.

File: Content.Server/Destructible/DestructibleSystem.cs (hunk 2 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:69]: This results in a testfail as it will cause a modified prototype on map save. This is why I had moved to OnMapInit previously.

Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:69]: Reverted to `OnMapInit` in d7394d31c02475b8dcf6b58628bf4732e2a08aa7

File: Content.Server/Destructible/DestructibleSystem.cs (hunk 3 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs]: Resolved in 900c8a7e33f48c0f729010e626d8dff95c8b6cc7 and 3d8b296776d10f35526af4d26d26f92a8b6fa0be
I used the Index built in method and rely on Insert to throw. 
One edge case I can see - if you index from the end too far and receive a negative index, it will wrap and insert from the end by that amount.

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 5 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Resolved in f9a9df9ac0cdc59a0d4afe4c4b0246daee01978b
Test is streamlined and more comprehensive.

File: Content.Server/Destructible/DestructibleComponent.cs [hunk 6]
```diff
@@ -1,25 +1,30 @@
...
 
-        /// <summary>
-        /// Specifies whether the entity has passed a damage threshold that causes it to break
-        /// </summary>
-        [DataField]
-        public bool IsBroken = false;
-    }
+    /// <summary>
+    /// Specifies if the entity should be silently destroyed on overkill damage.
+    /// </summary>
```
Comment (iaada) [inline Content.Server/Destructible/DestructibleComponent.cs:28]: Add a remark describing what "overkill damage" is.

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+using Content.Shared.Damage.Prototypes;
+using Content.Shared.Damage.Systems;
+using Content.Shared.Destructible;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Prototypes;
+using static Content.IntegrationTests.Tests.Destructible.DestructibleTestPrototypes;
+
+namespace Content.IntegrationTests.Tests.Destructible;
+
+public sealed class DestructibleOverkillTest
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: This test should inherit from `GameTest` to make use of the helper functions it gives. Also add a description of what this test tests.
```suggestion
public sealed class DestructibleOverkillTest : GameTest
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    public async Task Test()
+    {
+        await using var pair = await PoolManager.GetServerClient();
+        var server = pair.Server;
+
+        var testMap = await pair.CreateTestMap();
+
+        var sEntityManager = server.ResolveDependency<IEntityManager>();
+        var sPrototypeManager = server.ResolveDependency<IPrototypeManager>();
+        var sEntitySystemManager = server.ResolveDependency<IEntitySystemManager>();
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Instead of resolving these in the unit test, use `GameTest`'s sided dependencies in the test class itself. Additionally, `GameTest` provides several common dependencies on its own like `SEntMan` for `IEntityManager` and `SProtoMan` for `IPrototypeManager`.

```suggestion
    [SidedDependency(Side.Server)] private readonly IEntitySystemManager _sEntSysManager = null!;
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+        var baseEntityCount = sEntityManager.EntityCount;;
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Extra `;`.
```suggestion
        var baseEntityCount = _sEntSysManager.EntityCount;
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+public sealed class DestructibleOverkillTest
+{
+    [Test]
+    public async Task Test()
+    {
+        await using var pair = await PoolManager.GetServerClient();
+        var server = pair.Server;
+
+        var testMap = await pair.CreateTestMap();
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]:
```suggestion
        var testMap = await Pair.CreateTestMap();
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+            sDestructibleEntity = sEntityManager.SpawnEntity(DestructibleDestructionEntityId, coordinates);
+            sTestThresholdListenerSystem = sEntitySystemManager.GetEntitySystem<TestDestructibleListenerSystem>();
+            sTestThresholdListenerSystem.ThresholdsReached.Clear();
+        });
+
+        await server.WaitAssertion(() =>
+        {
+            var bruteDamageGroup = sPrototypeManager.Index<DamageGroupPrototype>(TestBruteDamageGroupId);
+            DamageSpecifier bruteDamage = new(bruteDamageGroup, 200);
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Define variables using `var` whenever possible.
```suggestion
            var bruteDamage = new DamageSpecifier(bruteDamageGroup, 200);
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        var baseEntityCount = sEntityManager.EntityCount;;
+
+        EntityUid sDestructibleEntity = default;
+        TestDestructibleListenerSystem sTestThresholdListenerSystem = null;
+
+        await server.WaitPost(() =>
+        {
+            var coordinates = testMap.GridCoords;
+
+            sDestructibleEntity = sEntityManager.SpawnEntity(DestructibleDestructionEntityId, coordinates);
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]:
```suggestion
            sDestructibleEntity = SSpawnAtPosition(DestructibleDestructionEntityId, coordinates);
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                sEntityManager.System<DamageableSystem>().TryChangeDamage(sDestructibleEntity, bruteDamage, true);
+            });
+
+            var threshold = sTestThresholdListenerSystem.ThresholdsReached[0].Threshold;
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(threshold.Triggered, Is.True);
+                Assert.That(threshold.Behaviors, Has.Count.EqualTo(1));
+            });
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: This test is missing comments, and here specifically is an important place to have them. What are you asserting, and how does this prove what you're testing?

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            var doActsBehavior = (DoActsBehavior)threshold.Behaviors.Single(b => b is DoActsBehavior);
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(doActsBehavior.HasAct(ThresholdActs.Destruction));
+            });
+        });
+
+        await server.WaitRunTicks(1);   // Wait for predicted delete
+        Assert.That(sEntityManager.EntityCount, Is.EqualTo(baseEntityCount), $"Overkill destructible test did not destroy cleanly.");
```
Comment (iaada) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: What does it mean to "not destroy cleanly" in this context? You should be explicit about what this asserts.

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 15]
```diff
@@ -19,210 +18,298 @@
...
+        {
+            entity.Comp.Thresholds.Add(threshold);
+        }
+    }
+
+    /// <summary>
+    /// Adds an overkill threshold if one does not exist.
+    /// </summary>
+    /// <param name="entity">The entity, component tuple to target.</param>
+    private void AddOverkillThreshold(Entity<DestructibleComponent> entity)
```
Comment (iaada) [inline Content.Server/Destructible/DestructibleSystem.cs:218]: Add a remark explaining what an "overkill threshold" is.

**Review by iaada — CHANGES_REQUESTED** [review]
There's been changes to how we'd like tests written, but the system changes look good to me.

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 12 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: b51914cf04869a068f66f611418a674fc98fb446

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 11 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 3c6c39cfb515b4ef86873811d3e4d9f80171c51e

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 10 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: b51914cf04869a068f66f611418a674fc98fb446

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 9 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: b51914cf04869a068f66f611418a674fc98fb446

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 8 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 3c6c39cfb515b4ef86873811d3e4d9f80171c51e

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 7 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: b51914cf04869a068f66f611418a674fc98fb446

File: Content.Server/Destructible/DestructibleComponent.cs (hunk 6 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleComponent.cs:28]: 791cd9c49c7c26b00831d44164c00fbb97281023

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 13 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: b51914cf04869a068f66f611418a674fc98fb446

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 14 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 3c6c39cfb515b4ef86873811d3e4d9f80171c51e

File: Content.Server/Destructible/DestructibleSystem.cs (hunk 15 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:218]: c7540e0122d7fab536183de27554fa0c3c191251

**Review by iaada — APPROVED** [review]
lgtm <3

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+using Content.Shared.Destructible;
+using Robust.Shared.GameObjects;
+using static Content.IntegrationTests.Tests.Destructible.DestructibleTestPrototypes;
+
+namespace Content.IntegrationTests.Tests.Destructible;
+
+/// <summary>
+/// Tests ensuring the correct operation of <see cref="SharedDestructibleSystem"/>.
+/// </summary>
+[TestFixture]
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]:
```suggestion
```
Not needed in modern NUnit versions

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        // Entity count prior to spawning and destroying
+        var baseEntityCount = SEntMan.EntityCount;
+
+        EntityUid sDestructibleEntity = default;
+        TestDestructibleListenerSystem sTestThresholdListenerSystem = null;
+
+        // Spawn our test entity and threshold listener
+        await Server.WaitPost(() =>
+        {
+            sDestructibleEntity = SEntMan.SpawnEntity(DestructibleDestructionEntityId, testMap.GridCoords);
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]:
```suggestion
            sDestructibleEntity = SSpawnAtPosition(DestructibleDestructionEntityId, testMap.GridCoords);
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    [Test]
+    public async Task EnsureOverkill()
+    {
+        var testMap = await Pair.CreateTestMap();
+
+        // Entity count prior to spawning and destroying
+        var baseEntityCount = SEntMan.EntityCount;
+
+        EntityUid sDestructibleEntity = default;
+        TestDestructibleListenerSystem sTestThresholdListenerSystem = null;
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: This should be another SidedDependency

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+        await Server.WaitAssertion(() =>
+        {
+            var bruteDamageGroup = SProtoMan.Index<DamageGroupPrototype>(TestBruteDamageGroupId);
+            var bruteDamage = new DamageSpecifier(bruteDamageGroup, 200);
+
+            // Hit the destructible with enough damage to overkill
+            Assert.DoesNotThrow(() =>
+            {
+                SEntMan.System<DamageableSystem>().TryChangeDamage(sDestructibleEntity, bruteDamage, true);
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: SidedDependency

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            });
+
+            // We now verify that our component has the properties we expect
+
+            // Our first threshold should be the overkill destruction
+            var threshold = sTestThresholdListenerSystem.ThresholdsReached[0].Threshold;
+
+            // Ensure that the threshold triggered and only has one behavior
+            Assert.Multiple(() =>
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Convert these to
```
using (Assert.EnterMultipleContext())
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+                Assert.That(threshold.Behaviors, Has.Count.EqualTo(1));
+            });
+
+            var doActsBehavior = (DoActsBehavior)threshold.Behaviors.Single(b => b is DoActsBehavior);
+
+            // Ensure that the one act in this behavior is destruction
+            Assert.Multiple(() =>
+            {
+                Assert.That(doActsBehavior.HasAct(ThresholdActs.Destruction));
+            });
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: Not a lot of point to a Multiple with only one Assert (I'm guessing you had more at one point).

File: Content.Server/Destructible/DestructibleComponent.cs [hunk 22]
```diff
@@ -1,25 +1,31 @@
...
-        /// </summary>
-        [DataField]
-        public bool IsBroken = false;
-    }
+    /// <summary>
+    /// Specifies if the entity should be silently destroyed when receiving damage significantly in excess of
+    /// its normal destructible threshold.
+    /// </summary>
+    [DataField(readOnly: true)]
+    public bool GenerateOverkillThreshhold = true;
```
Comment (Tayrtahn) [inline Content.Server/Destructible/DestructibleComponent.cs]: One h in threshold

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 23]
```diff
@@ -19,210 +18,302 @@
...
+
+[UsedImplicitly]
+public sealed partial class DestructibleSystem : SharedDestructibleSystem
 {
-    [UsedImplicitly]
-    public sealed partial class DestructibleSystem : SharedDestructibleSystem
+    [Dependency] public IAdminLogManager AdminLogger = default!;
+    [Dependency] public IPrototypeManager PrototypeManager = default!;
+    [Dependency] public IRobustRandom Random = default!;
+    public new IEntityManager EntityManager => base.EntityManager;
```
Comment (Tayrtahn) [inline Content.Server/Destructible/DestructibleSystem.cs:33]: I see this was already here, but what in the world? I would just delete this.

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 24]
```diff
@@ -19,210 +18,302 @@
...
+        AddOverkillThreshold(entity);
+    }
+
+    /// <summary>
+    /// Check if any thresholds were reached. if they were, execute them.
+    /// </summary>
+    private void OnDamageChanged(Entity<DestructibleComponent> entity, ref DamageChangedEvent args)
+    {
+        var component = entity.Comp;
+        var uid = entity.Owner;
```
Comment (Tayrtahn) [inline Content.Server/Destructible/DestructibleSystem.cs]:
```suggestion
        var (uid, component) = entity;
```
Optional suggestion.

**Review by Tayrtahn — CHANGES_REQUESTED** [review]

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 16 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 17 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 18 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 19 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 20 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs (hunk 21 above)
Comment (kontakt) [reply Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs]: 44cc6174eeca485c2956f15517f30b349107c999

Thanks for pointing that out, my eyes had glazed over at this point.

File: Content.Server/Destructible/DestructibleComponent.cs (hunk 22 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleComponent.cs]: 95ffd67ed29389e2074317acb14ed1b0ebedbba8

File: Content.Server/Destructible/DestructibleSystem.cs (hunk 23 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:33]: As you mentioned in discord, this is here to change access on `EntityManager` from  `Protected` to `Public`.

File: Content.Server/Destructible/DestructibleSystem.cs (hunk 24 above)
Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs]: 5a2b8e8e016df194a66becc4a527c3706cdf269e

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+/// </summary>
+public sealed class DestructibleOverkillTest : GameTest
+{
+    [SidedDependency(Side.Server)] private DamageableSystem _sDamageableSystem = default!;
+    [SidedDependency(Side.Server)] private TestDestructibleListenerSystem _sDestructibleListenerSystem = default!;
+
+    /// <summary>
+    /// Test that an entity with consequences is destroyed cleanly when overkilled.
+    /// </summary>
+    [Test]
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs:26]:
```suggestion
    [Test]
    [Description("Test that an entity with consequences is destroyed cleanly when overkilled.")]
```

**Review by Tayrtahn — APPROVED** [review]
Looks good to me!

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs:26]: Oh right, also
```suggestion
    [Test]
    [TestOf(typeof(DestructibleSystem))]
```

File: Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs [hunk 27]
```diff
@@ -0,0 +1,78 @@
+using System.Linq;
+using Content.IntegrationTests.Fixtures;
+using Content.IntegrationTests.Fixtures.Attributes;
+using Content.Server.Destructible.Thresholds.Behaviors;
```
Comment (Tayrtahn) [inline Content.IntegrationTests/Tests/Destructible/DestructibleOverkillTest.cs:5]:
```suggestion
using Content.IntegrationTests.Fixtures.Attributes;
using Content.Server.Destructible;
using Content.Server.Destructible.Thresholds.Behaviors;
```

**Review by Tayrtahn — COMMENTED** [review]
Guh. Sorry.

File: Content.Server/Destructible/DestructibleSystem.cs [hunk 28]
```diff
@@ -19,210 +18,301 @@
...
+            Trigger = new DamageTrigger { Damage = FixedPoint2.Max(MinimumOverkill, OverkillMultiplier * maxTrigger) },
+            Behaviors = { new DoActsBehavior { Acts = ThresholdActs.Destruction } },
+        };
 
-            destroyedAt = DestroyedAt(ent, ent.Comp);
-            return true;
-        }
+        // Thresholds are evaluated in order, so overkill must be first to avoid triggering effects
+        AddThreshold(entity.AsNullable(), autoThreshold, 0);
+    }
```
Comment (Princess-Cheeseballs) [inline Content.Server/Destructible/DestructibleSystem.cs:256]: A bit late to request changes but I'm not a huge fan of this logic. It would probably be smarter to query the objects it destructs into and generate an overkill threshold based on that. This would be ideal for explosions to prevent us from having to do logic to overflow damage onto entities created by an object breaking, by baking that behavior into DestructibleSystem itself.

Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: I don't follow.
Right now the destructible threshold added by the overkill is placed at the lowest index. That means it will execute first and then short circuit. This prevents spawning entities that require processing at all. 
The overkill also prevents any non-spawning effects that could be associated with a breakage event, e.g. sounds, from causing a lot of simultaneous activities.

Comment (Princess-Cheeseballs) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: Effectively rather than adding a new trigger that's a multiplier of the regular spawn entities trigger, we check the spawn entities trigger, then check which entities it breaks down into and see how much damage is needed to destroy those entities, and set the overkill threshold to the additive damage value of breaking this entity plus those entities. 

It wouldn't be perfect but it would make a lot more sense, especially for structures like walls where the girders are extremely weak compared to the wall above them.

Comment (whatston3) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: > It wouldn't be perfect but it would make a lot more sense, especially for structures like walls where the girders are extremely weak compared to the wall above them.

Perhaps it would be more straightforward for those cases to have explicit [overkill] thresholds in YAML, then.  Additionally, it becomes easier to lose out on any feedback (e.g. sounds playing) if the overkill threshold is very close to your normal destructive behaviour.

Is it worth introducing recursive complexity for something that could just be specified in special cases?

Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: Most things decompose into, like, a sheet of steel.
I could add logic that checks the destructible and recurses into its results and fails out to the 2x if there are no nested destructibles if you want. I don't know if that additional code complexity buys us much though.

Comment (Princess-Cheeseballs) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: Actually, speaking of that while this does only run on mapinit, it runs on mapinit for all destructible entities. The spawning of these entities is already somewhat of a bottleneck for performance so I'm not even sure we should be doing these kinds of calculations.

Ideally we should just be autogenerating this kind of YAML into the prototype rather than editing each entity on spawn. Did you run a before and after Destructible Benchmark for this PR?

Comment (kontakt) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: So I compared the before and after of this PR with the destructible benchmark
Before:
image

After:
image

That last run failed with a PVS entity error - I can rerun it if you want but it's a real slow one.
Outcome is a wash, the measurements are pretty noisy and there's no big difference. Mostly because spawning the entities is done in setup and not in the timed loop.

Testing map loading benchmark:
Before:
image

After:
image

Again, no result that points to a heavy impact with this PR.

I originally used `OnComponentInit` but that trips the modified prototype test. The point of doing this was to *reduce* the amount of boilerplate YAML and make behavior more consistent.

Comment (Princess-Cheeseballs) [reply Content.Server/Destructible/DestructibleSystem.cs:256]: Alright that's good to hear.
